src/typecheck.cpp

```diff
@@ -236,13 +237,39 @@ void typecheck(TypedAST::RecordAccessExpression* ast, TypeChecker& tc) {
 	MonoId member_type = tc.new_var();
 	ast->m_value_type = member_type;
 
-	TypeFunctionId dummy_tf = tc.m_core.new_dummy_type_function(
-	    TypeFunctionTag::Record, {{ast->m_member->m_text.str(), member_type}});
+	TypeFunctionId dummy_tf = tc.m_core.new_type_function
+	    ( TypeFunctionTag::Record
+	    , {ast->m_member->m_text.str()}
+	    , {{ast->m_member->m_text.str(), member_type}}
+	    , true);
 	MonoId term_type = tc.m_core.new_term(dummy_tf, {}, "record instance");
 
 	tc.m_core.m_mono_core.unify(ast->m_record->m_value_type, term_type);
 }
 
+void typecheck(TypedAST::ConstructorExpression* ast, TypeChecker& tc) {
+	typecheck(ast->m_constructor, tc);
+	TypeFunctionId tf = tc.m_core.m_tf_core.find_term(ast->m_constructor->m_value_type);
+	TypeFunctionData& tf_data = tc.m_core.m_type_functions[tf];
+
+	// TODO: match type arguments as well, we only support 0 now
+
+	// match value arguments
+	assert(tf_data.fields.size() == ast->m_args.size());
+	for (int i = 0; i < ast->m_args.size(); ++i) {
+		typecheck(ast->m_args[i], tc);
+
+		MonoId field_type = tf_data.structure[tf_data.fields[i]];
+		tc.m_core.m_mono_core.unify(field_type, ast->m_args[i]->m_value_type);
+	}
+
+	// NOTE: this is a special case for the `name { }` syntax sugar,
+	// but perhaps the left side should only be a TypeTerm and not possibly an Identifier?
+	if (ast->m_constructor->m_meta_type == tc.meta_typefunc())
+		ast->m_value_type =
+		    tc.m_core.new_term(ast->m_constructor->m_value_type, {}, "record instance");
```

Comment:
Also, once we reach `typecheck`, there should be no `TypedAST::TypeTerm`s anymore, they should all have been turned into `MonoTypeHandle`s

Reply:
I was wrong on this comment, BTW.

This part in particular:

> so is assuming that if the metatype is typefunc, then the node is an identifier (e.g. it could be a StructExpression).

It could only be either a TypeFunctionHandle, or an Identifier (Only a Handle with my latest patch)